### ISSUE TITLE
managing canvases properly ...

### ISSUE THREAD

**Issue by mmeeks** [CollaboraOnline/online#6369]
Since we switched to rendering in JS canvases in 22.05 - so that we can apply deltas easily to them, we have an issue around lifecycle managing them at least on iOS - and for large documents where we don't get any giant invalidations to clear our cache out (as we tend to do while editing).

There is a risk that as we reduce giant invalidations we start to gum up our browsers more and more with canvases - which are often a scarce (GPU) resource. Patrick reports there is a 225Mb limit on Mac for canvas sizes, after which we just stop working. Worse canvases are GC's at a time of the browsers' choosing not ours. However - we can release the memory back by manually setting the size to nothing (fun) =)

```
commit f2f0cd10e10c63aa41715b7b5259cc93ccb432c9
Author: Patrick Luby <patrick.luby@collabora.com>
Date:   Thu Apr 20 19:34:37 2023 -0400

    Partial fix for issue #5876 discard excess canvas contexts
    
    WKWebView has a hardcoded memory limit for all canvas contexts
    so if canvas.getContext('2d') returns null, convert the canvas
    of other tiles to an image until canvas.getContext('2d') succeeds.
   
``` 

Should probably be re-worked to ensure that when life is calm (how to detect that is left as an exercise for the reader =) - and we are running lowish on canvas memory (that needs accounting for I guess) that we re-compress all tiles that are canvas' into these horror PNG images - if they are far away from our current view.

We should prolly poke: browser/src/layer/tile/CanvasTileLayer.js - where the tilecache is this._tileCache[] to clean this up sensibly before we start taking desperate measures at the point that we allocate a canvas.

Thanks ! =)

**Comment by mmeeks** [comment]
See also: #5876

**Comment by plubius** [comment]
I think you are looking for the following 23.05 patch:

https://github.com/CollaboraOnline/online/pull/6375/commits/81773510ef7c78021e5874cf2b2b170d83c2a833

If you replace the following line:
    if (!ctx && window.ThisIsTheiOSApp) {
with:
    if (!ctx) {
maybe that fixes the problem in Linux browsers?

**Comment by mmeeks** [comment]
A start of this is here: https://github.com/CollaboraOnline/online/compare/private/mmeeks/canvasbits?expand=1
As yet un-tested even by me; but thoughts appreciated.

**Comment by plubius** [comment]
iOS now builds so I'll test your code on my Mac tomorrow. Interesting that you hit a limit on Chrome and/or Firefox browsers. I wonder if this is a new thing.

**Comment by mmeeks** [comment]
Wow -I'm pulling a long string here that turns into a chain I'm afraid ... this is not going to be done today ;-)

**Comment by plubius** [comment]
I'll give this a test on iOS with the .pdf in #5876. On iOS, it is fairly easy to hit the maximum canvas memory limit after scrolling through that file.

**Comment by plubius** [comment]
> I'll give this a test on iOS with the .pdf in issue #5876. On iOS, it is fairly easy to hit the maximum canvas memory limit after scrolling through that file.

So far this looks like it is working well. Very responsive drawing even after zooming in and loading most of the pages of the PDF and I have yet to see the "out of canvas memory" console warnings in Safari's Web Inspector.

I did see two small things:
- There was a recurring type error logged in the Safari Web Inspector which stopped with [this patch](https://github.com/CollaboraOnline/online/files/11807344/canvastilelayerpatch.txt)
- On my basic model iPad and even with release build settings, the WKWebView's rendering subprocess will crash if you zoom in and scroll through the .pdf in issue #5876 long enough. I am receiving iOS native memory warnings just before the crash.

For the memory warnings, I assume that I just need to call _garbageCollect(). If so, I can try to wire that up to the iOS app later this week.

**Comment by mmeeks** [comment]
Caolan and I are working on the tilewins branch here and he just pushed that fix - thanks =)

**Comment by mmeeks** [comment]
@plubius can we get some more details on what the memory usage is like before the crash? possibly we need to tweak the garbage collection there somehow (?) would be nice to see what a memory profile looks like just beforehand ;-) do we need to lower a high-watermark in _garbageCollect - or (perhaps better)  - adapt the watermarks based on the actual screen-size (which tends to be many fewer tiles on mobile vs. a 4k screen eg. =) prolly something calculated there rather than hard-coded would be more effective. Anyhow - thanks for the testing & patch =)

**Comment by plubius** [comment]
So I was able to attach the "com.apple.Webkit.WebContent" processes on the iPad and the combined memory (see screen snapshot below) just kept growing to about 1.1 GB right below the crash. If I can implement a garbage collect call from the native code to JS, I'll post what I see after any memory warnings are received.
<img width="1396" alt="Screen Shot 2023-06-21 at 1 52 00 PM" src="https://github.com/CollaboraOnline/online/assets/104370511/1b357704-6adb-431c-9009-15f2eb08f8e6">

**Comment by mmeeks** [comment]
Interesting =) attaching a remote chrome or safari to the webkit thing should let us see where the memory is going - but will check locally too.

**Comment by plubius** [comment]
So I added a console.warn() in _garbageCollect and I am seeing it run frequently so adding a native to JS low memory call likely won't make any different. Are there any value changes that you think will have the most effect?

**Comment by mmeeks** [comment]
If memory grows un-controlledly; we should do a JS heap profile to see what is going on there - that should be possible in most remote debugging tools. Doing one here.

**Comment by mmeeks** [comment]
But don't see much that's interesting in the heap profile even after a lot of slide flipping here at least. I wonder if we're not GC'ing canvases properly - and whether (perhaps) we should not set their size to 0 - but have a small pool of these that we just re-use (?)

**Comment by plubius** [comment]
[Heap Snapshot 2023-06-21 at 14.50.35.json.txt](https://github.com/CollaboraOnline/online/files/11823770/Heap.Snapshot.2023-06-21.at.14.50.35.json.txt)

Uploaded my heap analysis file from Safari. If you can't read it see the screen snapshot below. the retained size appears to be from some _logs array holding on to a huge number of string messages.<img width="1343" alt="Screen Shot 2023-06-21 at 2 55 39 PM" src="https://github.com/CollaboraOnline/online/assets/104370511/a95d96e5-241d-420d-8924-106a9e4d887e">

**Comment by plubius** [comment]
Found the source of the endlessly growing _logs array: its _fancyLog() in browser/src/layer/marker/TextInput.js:

        // Save to downloadable log
        L.Log.log(payload.toString(), 'INPUT');

**Comment by plubius** [comment]
There is one more in _onTileMsg() in browser/src/layer/tile/CanvasTileLayer.js :

            L.Log.log(textMsg, 'INCOMING', key);

 This is the really big source of retained memory than the log() call in my previous post as textMsg is 256K on iOS as it holds the hex encoded image data in the tile messages.

**Comment by mmeeks** [comment]
Riight; that's some very old code from Mihai Varga - we should hack loleaflet/src/core/Log.js to only keep the last 100 lines or something I think; and (I guess) this is particularly problematic for iOS & Android when those input strings are super-long in text format and contain the whole compressed tiles; perhaps we should crop those 128 characters in or something as well ? in that payload.toString().

Any chance of a cross-platform fix for both of those ? - and really glad it's not the tile stuff that's burning RAM this once =)

**Comment by plubius** [comment]
> Any chance of a cross-platform fix for both of those ? - and really glad it's not the tile stuff that's burning RAM this once =)

I think I can whip something up this week. 128 characters and limit Logs._logs array to 100 seems reasonable.

**Comment by plubius** [comment]
OK. Below is a quick and dirty patch the limits the message size and number of entries in the Log class. If shift() might be a performance hit, is there a queue implementation somewhere in the JS code?:

[logmemorylimits.diff.txt](https://github.com/CollaboraOnline/online/files/11826634/logmemorylimits.diff.txt)

Anyways, max memory usage by the WKWebView subprocess is significantly lower with the above patch. It never exceeds 0.5 GB now compared to 1.1 GB before. Also, below is a heap analysis with the patch. The retained memory at the top of the list is, AFAICT, only HTMLCanvasElement instances and the rawData data member in tile instances:

[Heap Snapshot 2023-06-21 at 18.12.36.json.txt](https://github.com/CollaboraOnline/online/files/11826630/Heap.Snapshot.2023-06-21.at.18.12.36.json.txt)

**Comment by mmeeks** [comment]
Patch looks golden - it's a great catch: https://github.com/CollaboraOnline/online/pull/6696 just needs your sign-off to get merged =) We expect to have the rawData members close to the top - although controlled by the limits in the GC code. Nice work.

**Comment by gitgt** [comment]
Can someone explain the status here, is there something that remains?

**Comment by plubius** [comment]
> Can someone explain the status here, is there something that remains?

IIRC, this is now fixed. But I only tested in the iOS app (i.e. with Safari's embedded WKWebView browser) so I don't know if there are still any unresolved issues when running Online in a regular web browser.

**Comment by mmeeks** [comment]
Lots more wins around canvas memory usage merged for 23.05 - we also found we were rendering to redundant off-screen canvases as well in TilesSection.ts - now gone =) Thanks !.
